### ISSUE TITLE
Can we please bring back method overload AcquireTokenAsync(string, string, UserCredential)?

### ISSUE THREAD

**Issue by TimLovellSmith** [AzureAD/azure-activedirectory-library-for-dotnet#1085]
Now when I try to use recent Microsoft Azure Rest SDK versions with higher versions of ADAL, like the latest, 3.19.6 I get this compatibility error.

"Method not found: 'System.Threading.Tasks.Task`1<Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationResult> Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext.AcquireTokenAsync(System.String, System.String, Microsoft.IdentityModel.Clients.ActiveDirectory.UserCredential)'."

These binary breaking changes are very painful to work through!

**Comment by jmprieur** [comment]
hello, @TimLovellSmith : 
- what was the previous version you were using? would it be ADAL v2.x ?
- The method is still there (as an extension method): https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/62a82c78b2a8504065f342309f8015dde8b9eccc/adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Features/WinCommon/AuthenticationContextIntegratedAuthExtensions.cs#L48
- It's available in .NET 4.5 and .NET Standard 1.3

Also which platform are you targeting (Xamarin.Android?). We've seen 2 cases where NuGet restore (the NuGet UI through Visual Studio, actually) did not restore the right platform.

cc: @henrik-me @bgavrilMS @jennyf19

**Comment by TimLovellSmith** [comment]
@jmprieur Unfortunately, being there as an extension method just isn't binary compatible, and the call to it is not from our code, it is from a DLL we depend on. An extension method is basically syntactic sugar for calling a static method, where the static method is on a different class. When you look at the compiled IL it is trying to call an instance method on the AuthenticationContext class, not a static method.
I'm targeting Windows desktop .net 4.6.2.

**Comment by jmprieur** [comment]
@TimLovellSmith 
you should expect breaking changes when moving from ADAL 2.x to ADAL 3.x as the major number was incremented. ADAL uses semantinc versioning (See https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Semantic-versioning.-API-change-management)

I agree that changing a method to an extension method is a binary breaking change. (I wonder why this was done ...)

Can you tell us more about your scenario? 
Do I understand correctly that your product leverages a library A, which itself is using ADAL 2.x?
and you want to reference ADAL 3.x, whereas library A has not upgraded ?

**Comment by TimLovellSmith** [comment]
@jmprieur 
I agree that you are doing semver right since you incremented the major version when you made breaking chagnes...

However! I don't think that is a great reason to make breaking changes. So got another reason?

In particular regarding binary compatibility I think you are better off trying to keep binary compatibility where possible, especially compatibility with other popular libraries (like Azure management). The reason is that in practice, getting all your nuget package versions to align so that everything just works gets very, very much harder once you have these binary incompatible changes (it can feel like waiting for the stars to come into alignment...)

E.g. a 'typical' scenarios of nuget package installation: if I create a new project, and I predict I am going to want to install ADAL, then nuget installs the latest version of ADAL for me by default. If I then go and install some other project which uses ADAL, nuget will generally *not* automatically downgrade ADAL (especially not if that project assumed it was binary compatible past the next major version, which while probably incorrect, is what I think frequently happens...) and everything builds... but then fails after you run it.

In short, there is a giant pit of nuget package failure to fall into here.

**Comment by jmprieur** [comment]
@TimLovellSmith , I was not part of the team when it happened, and therefore I've no idea why this binary breaking change was made, but it was made between 2.x and 3.x. We cannot undo this.
ADAL 3.x hasGAed on May 11th 2016, that is more than 2 years ago (See https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/releases/tag/v3.10.305110106), with many (minor) versions since. Your PR would binary break all these versions (including VS / VSTS interestingly)

The right solution is for us to work with Azure management so that they move to ADAL 3.x, or even maybe MSAL.NET. Can you please point me to the NuGet package that you consume?

Also I'm still interested in understanding what your scenario is. Why do you get this error if you reference Azure management only ? Surely you reference ADAL 3.x yourself? or through another library? Can you please tell us more?

**Comment by TimLovellSmith** [comment]
>Your PR would binary break all these versions (including VS / VSTS interestingly)

@jmprieur I don't think it would, but maybe I'm overlooking something - whats the breaking change I'm making?

**Comment by bgavrilMS** [comment]
I checked and your fix does indeed seems to not break binary compatibility. 

But be advised that all the assemblies are signed / strongly named, and as such you won't be able to just copy a new assembly and over the old one. Binary versioning is enforced by strong naming. You'd have add some binding redirects over your solution or do something along the lines of https://stackoverflow.com/questions/3158928/referencing-2-different-versions-of-log4net-in-the-same-solution. 

We'd also have to release another version of ADAL3, which is not trivial, especially as this issue doesn't seem blocking and the "fix" is likely more work than doing the right thing and updating your dependency. 

@jmprieur - do we have guidelines for upgrading from ADAL v2 to ADAL v3 ?

**Comment by jmprieur** [comment]
@bgavrilMS : no we don't yet. I guess it would be a good idea to add an article in the conceptual documentation. Added a story for that in the backlog

**Comment by TimLovellSmith** [comment]
@bgavrilMS Well, personally I think it could be pretty cool if you could put out another release of ADAL v3 with this kind of binary compat fix - I guess there may be other issues like this, this is just the first one I happened to run into. I have been trying to get my app to work with ADAL v2 and unfortunately all I could get so far was 500 internal server error to come back from the server, no idea why. ;-(

**Comment by bgavrilMS** [comment]
Might as well upgrade to ADAL v3 then? v2 is quite old and you'll have trouble finding someone to help you troubleshoot, even in the ADAL team. 

Also note that apart for binary compatibility, v2 and v3 are not even logically compatible - constant names like error codes have changed etc. 

I thinnk we're not going to take this fix as we cannot ensure full compatibility between v2 and v3. I would recommend you either upgrade to v3 or you follow this workaround for using 2 versions of the same library in your code: https://stackoverflow.com/questions/3158928/referencing-2-different-versions-of-log4net-in-the-same-solution
